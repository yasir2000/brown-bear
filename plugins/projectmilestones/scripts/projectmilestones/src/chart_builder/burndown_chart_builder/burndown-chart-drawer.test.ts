/*
 * Copyright (c) Enalean, 2019 - present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */

import { createBurndownChart, getMaxRemainingEffort } from "./burndown-chart-drawer";

import type { BurndownData, PointsWithDateForBurndown } from "../../type";
import type { ChartPropsWhithoutTooltip } from "../../../../../../../src/scripts/charts-builders/type";
import * as chart_badge_generator from "../chart-badge-generator";

jest.mock("../../../../../../../src/scripts/charts-builders/time-scale-labels-formatter");
jest.mock("../time-scale-label-formatter");
jest.mock("../chart-badge-generator", () => ({
    addBadgeCaption: jest.fn(),
}));

describe("BurndownChartDrawer -", () => {
    beforeEach(() => jest.clearAllMocks());

    describe("getMaxRemainingEffort -", () => {
        it("Returns the highest remaining effort if it is greater than the capacity", () => {
            const max_remaining_effort = getMaxRemainingEffort({
                points_with_date: getPointsWithDateWithMaxIs15(),
                capacity: 10,
            } as BurndownData);

            expect(max_remaining_effort).toEqual(15);
        });

        it("Returns capacity if it is greater than the biggest remaining effort", () => {
            const max_remaining_effort = getMaxRemainingEffort({
                points_with_date: getPointsWithDateWithMaxIs15(),
                capacity: 20,
            } as BurndownData);

            expect(max_remaining_effort).toEqual(20);
        });

        it("When there are some remaining effort and no capacity, Then the biggest remaining effort is returned", () => {
            const max_remaining_effort = getMaxRemainingEffort({
                points_with_date: getPointsWithDateWithMaxIs15(),
                capacity: null,
            } as BurndownData);

            expect(max_remaining_effort).toEqual(15);
        });

        it("When there aren't capacity and remaining effort, Then 5 is returned", () => {
            const max_remaining_effort = getMaxRemainingEffort({
                points_with_date: getPointsWithDateWithMaxIsNull(),
                capacity: null,
            } as BurndownData);

            expect(max_remaining_effort).toEqual(5);
        });

        it("When there aren't remaining effort but there is capacity, Then capacity is returned", () => {
            const max_remaining_effort = getMaxRemainingEffort({
                points_with_date: getPointsWithDateWithMaxIsNull(),
                capacity: 100,
            } as BurndownData);

            expect(max_remaining_effort).toEqual(100);
        });

        it("When capacity and remaining effort are 0, Then 5 is returned", () => {
            const max_remaining_effort = getMaxRemainingEffort({
                points_with_date: getPointsWithDateWithMaxIsZero(),
                capacity: 0,
            } as BurndownData);

            expect(max_remaining_effort).toEqual(5);
        });
    });

    describe("createBurndownChart -", () => {
        it("When the chart is created, Then there are a G element and 2 lines scale and curve", () => {
            const chart_svg_element = getDocument();
            createBurndownChart(chart_svg_element, getChartProps(), getBurndownData(), 101);

            expect(chart_svg_element).toMatchSnapshot();
            expect(chart_badge_generator.addBadgeCaption).toBeCalled();
        });

        it("When there isn't points with date in burndownData, Then there is an empty graph", () => {
            const chart_svg_element = getDocument();
            createBurndownChart(
                chart_svg_element,
                getChartProps(),
                getBurndownDataWithoutPointsWithDate(),
                101
            );

            expect(chart_badge_generator.addBadgeCaption).not.toBeCalled();
            expect(chart_svg_element.childElementCount).toEqual(1);
        });

        it("When the last point has 0 remaining effort, Then there isn't badge", () => {
            const chart_svg_element = getDocument();
            createBurndownChart(
                chart_svg_element,
                getChartProps(),
                getBurndownDataWith0RemainingEffort(),
                101
            );

            expect(chart_badge_generator.addBadgeCaption).not.toBeCalled();
            expect(chart_svg_element.childElementCount).toEqual(1);
        });

        function getDocument(): HTMLElement {
            const local_document = document.implementation.createHTMLDocument();
            const chart_div = local_document.createElement("svg");
            chart_div.setAttribute("id", "chart-100");
            return chart_div;
        }
    });

    function getChartProps(): ChartPropsWhithoutTooltip {
        return {
            graph_width: 100,
            graph_height: 100,
            margins: {
                top: 20,
                right: 20,
                bottom: 20,
                left: 20,
            },
        };
    }

    function getBurndownData(): BurndownData {
        return {
            opening_days: [1, 2, 3, 4, 5],
            duration: 1,
            start_date: "2019-07-01T00:00:00+00:00",
            capacity: null,
            is_under_calculation: false,
            points: [5, 10, 15],
            points_with_date: getPointsWithDateWithMaxIs15(),
        } as BurndownData;
    }

    function getBurndownDataWithoutPointsWithDate(): BurndownData {
        return {
            opening_days: [1, 2, 3, 4, 5],
            duration: 1,
            start_date: "2019-07-01T00:00:00+00:00",
            capacity: null,
            is_under_calculation: false,
            points: [] as number[],
            points_with_date: [] as PointsWithDateForBurndown[],
        } as BurndownData;
    }

    function getBurndownDataWith0RemainingEffort(): BurndownData {
        return {
            opening_days: [1, 2, 3, 4, 5],
            duration: 1,
            start_date: "2019-07-01T00:00:00+00:00",
            capacity: null,
            is_under_calculation: false,
            points: [0],
            points_with_date: getPointsWithDateWithMaxIsZero(),
        } as BurndownData;
    }

    function getPointsWithDateWithMaxIs15(): PointsWithDateForBurndown[] {
        const points: PointsWithDateForBurndown[] = [];
        points.push({
            date: "2019-07-01T00:00:00+00:00",
            remaining_effort: null,
        });
        points.push({
            date: "2019-07-01T00:00:00+00:00",
            remaining_effort: 15,
        });

        return points;
    }

    function getPointsWithDateWithMaxIsNull(): PointsWithDateForBurndown[] {
        const points: PointsWithDateForBurndown[] = [];
        points.push({
            date: "2019-07-01T00:00:00+00:00",
            remaining_effort: null,
        });

        return points;
    }

    function getPointsWithDateWithMaxIsZero(): PointsWithDateForBurndown[] {
        const points: PointsWithDateForBurndown[] = [];
        points.push({
            date: "2019-07-01T00:00:00+00:00",
            remaining_effort: 0,
        });

        return points;
    }
});
