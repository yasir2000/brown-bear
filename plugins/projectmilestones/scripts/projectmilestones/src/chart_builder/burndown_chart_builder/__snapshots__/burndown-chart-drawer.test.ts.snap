// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BurndownChartDrawer - createBurndownChart - When the chart is created, Then there are a G element and 2 lines scale and curve 1`] = `
<svg
  id="chart-100"
>
  <svg
    height="100"
    width="100"
  >
    <g
      class="chart-x-axis"
      fill="none"
      font-family="sans-serif"
      font-size="10"
      text-anchor="middle"
      transform="translate(0, 80)"
    >
      <g
        class="tick"
        opacity="1"
        transform="translate(20.5,0)"
      >
        <text
          dy="0.71em"
          fill="currentColor"
          y="11"
        >
          2019-07-01
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(80.5,0)"
      >
        <text
          dy="0.71em"
          fill="currentColor"
          y="11"
        >
          2019-07-02
        </text>
      </g>
    </g>
    <g
      class="chart-y-axis"
      fill="none"
      font-family="sans-serif"
      font-size="10"
      text-anchor="end"
      transform="translate(20, 0)"
    >
      <g
        class="tick"
        opacity="1"
        transform="translate(0,80.5)"
      >
        <text
          dy="0.32em"
          fill="currentColor"
          x="-5"
        >
          0
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,61.75)"
      >
        <text
          dy="0.32em"
          fill="currentColor"
          x="-5"
        >
          5
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,43)"
      >
        <text
          dy="0.32em"
          fill="currentColor"
          x="-5"
        >
          10
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,24.25)"
      >
        <text
          dy="0.32em"
          fill="currentColor"
          x="-5"
        >
          15
        </text>
      </g>
    </g>
    <g
      class="ideal-line"
    >
      <path
        d="M20,23.75L80,80"
      />
    </g>
    <line
      class="release-line-scale"
      x1="20"
      x2="90"
      y1="80"
      y2="80"
    />
    <line
      class="release-line-scale"
      x1="20"
      x2="20"
      y1="80"
      y2="13.75"
    />
    <path
      class="chart-curve-remaining-effort"
      d="M20,23.75Z"
    />
  </svg>
</svg>
`;
